File: xplatform/lib/import/mobi_importer.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import '../alignment/alignment_types.dart';
import 'ebook_importer.dart';

// References:
//   - https://wiki.mobileread.com/wiki/MOBI
//   - https://wiki.mobileread.com/wiki/PDB
// Only the MOBI 6 / PalmDOC subset is implemented; KF8 (.azw3) and
// HUFF/CDIC throw rather than silently producing garbage.
class MOBIImporter implements EBookImporter {
  const MOBIImporter();

  @override
  Future<ImportedBook> importBook(File file) async =>
      _decode(await file.readAsBytes());

  ImportedBook _decode(Uint8List bytes) {
    final db = _PalmDB.parse(bytes);
    if (db.type != 'BOOK' || db.creator != 'MOBI') {
      throw ImporterError(
          'Not a MOBI file (PDB type="${db.type}", creator="${db.creator}").');
    }
    if (db.recordCount == 0) {
      throw const ImporterError('MOBI file has no records.');
    }

    final rec0 = db.recordBytes(0);
    final pdoc = _PalmDOCHeader.parse(rec0);
    if (pdoc.encryption != 0) {
      throw const ImporterError('DRM-protected MOBI; cannot import.');
    }
    final mobi = _MOBIHeader.parse(rec0);
    if (pdoc.compression == _Compression.huffCdic) {
      throw const ImporterError(
          'HUFF/CDIC compression unsupported. Convert to EPUB with Calibre.');
    }
    if (mobi.isKF8) {
      throw const ImporterError(
          '.azw3 / KF8 unsupported. Convert to EPUB with Calibre.');
    }

    final html = _readBookText(db, pdoc, mobi);
    final exth = _EXTHData.parse(rec0, mobi);
    final title =
        (exth.updatedTitle ?? _readFullName(rec0, mobi) ?? db.name).trim();
    final cover = _findCover(db, mobi, exth);

    return ImportedBook(
      title: title.isEmpty ? db.name : title,
      author: exth.author ?? 'Unknown',
      coverImageData: cover,
      segments: _splitOnHeadings(html),
    );
  }
}

String _readBookText(_PalmDB db, _PalmDOCHeader pdoc, _MOBIHeader mobi) {
  final lastText = mobi.firstNonBookIndex > 1
      ? mobi.firstNonBookIndex - 1
      : pdoc.recordCount;
  final buf = BytesBuilder();
  for (var i = 1; i <= lastText && i < db.recordCount; i++) {
    final raw = _stripTrailers(db.recordBytes(i), mobi.extraDataFlags);
    buf.add(pdoc.compression == _Compression.palmDoc
        ? _palmdocInflate(raw)
        : raw);
  }
  return _decodeBytes(buf.toBytes(), mobi.textEncoding);
}

Uint8List? _findCover(_PalmDB db, _MOBIHeader mobi, _EXTHData exth) {
  final fromExth =
      exth.coverIndex != null ? _imageAt(db, mobi, exth.coverIndex!) : null;
  return fromExth ?? _firstImage(db, mobi);
}

abstract class _Compression {
  static const palmDoc = 2;
  static const huffCdic = 17480;
}

String _decodeBytes(Uint8List bytes, int encoding) =>
    encoding == 65001 // UTF-8; otherwise the only seen value is 1252 (cp1252).
        ? utf8.decode(bytes, allowMalformed: true)
        : latin1.decode(bytes);

class _PalmDB {
  _PalmDB._(
      this.name, this.type, this.creator, this.recordOffsets, this.bytes);

  final String name;
  final String type;
  final String creator;
  final List<int> recordOffsets;
  final Uint8List bytes;

  int get recordCount => recordOffsets.length;

  static _PalmDB parse(Uint8List bytes) {
    if (bytes.length < _Pdb.headerSize) {
      throw const ImporterError('File too short to be a PalmDB.');
    }
    final bd = ByteData.sublistView(bytes);
    final count = bd.getUint16(_Pdb.recordCount, Endian.big);
    if (_Pdb.headerSize + count * _Pdb.recordEntrySize > bytes.length) {
      throw const ImporterError('PalmDB record table truncated.');
    }
    final offsets = [
      for (var i = 0; i < count; i++)
        bd.getUint32(_Pdb.headerSize + i * _Pdb.recordEntrySize, Endian.big),
    ];
    return _PalmDB._(
      _readCString(bytes, _Pdb.name, _Pdb.nameLength),
      latin1.decode(bytes.sublist(_Pdb.type, _Pdb.type + 4)),
      latin1.decode(bytes.sublist(_Pdb.creator, _Pdb.creator + 4)),
      offsets,
      bytes,
    );
  }

  Uint8List recordBytes(int idx) {
    final start = recordOffsets[idx];
    final end =
        idx + 1 < recordOffsets.length ? recordOffsets[idx + 1] : bytes.length;
    return bytes.sublist(start, end);
  }
}

abstract class _Pdb {
  static const headerSize = 78;
  static const recordEntrySize = 8;
  static const name = 0;
  static const nameLength = 32;
  static const type = 60;
  static const creator = 64;
  static const recordCount = 76;
}

String _readCString(Uint8List bytes, int offset, int maxLen) {
  var stop = offset + maxLen;
  for (var i = offset; i < offset + maxLen; i++) {
    if (bytes[i] == 0) {
      stop = i;
      break;
    }
  }
  return latin1.decode(bytes.sublist(offset, stop));
}

class _PalmDOCHeader {
  _PalmDOCHeader(this.compression, this.recordCount, this.encryption);

  final int compression;
  final int recordCount;
  final int encryption;

  static _PalmDOCHeader parse(Uint8List rec0) {
    if (rec0.length < 16) {
      throw const ImporterError('Record 0 too short.');
    }
    final bd = ByteData.sublistView(rec0);
    return _PalmDOCHeader(
      bd.getUint16(0, Endian.big),
      bd.getUint16(8, Endian.big),
      bd.getUint16(12, Endian.big),
    );
  }
}

class _MOBIHeader {
  _MOBIHeader({
    required this.headerLength,
    required this.textEncoding,
    required this.firstNonBookIndex,
    required this.fullNameOffset,
    required this.fullNameLength,
    required this.firstImageRecord,
    required this.exthFlags,
    required this.extraDataFlags,
    required this.isKF8,
  });

  final int headerLength;
  final int textEncoding;
  final int firstNonBookIndex;
  final int fullNameOffset;
  final int fullNameLength;
  final int firstImageRecord;
  final int exthFlags;
  final int extraDataFlags;
  final bool isKF8;

  bool get hasEXTH => (exthFlags & 0x40) != 0;

  static _MOBIHeader parse(Uint8List rec0) {
    if (rec0.length < _Mobi.minLength ||
        latin1.decode(rec0.sublist(_Mobi.magic, _Mobi.magic + 4)) != 'MOBI') {
      throw const ImporterError('MOBI header not found in record 0.');
    }
    final bd = ByteData.sublistView(rec0);
    int u32(int o) => o + 4 <= rec0.length ? bd.getUint32(o, Endian.big) : 0;
    final mobiType = u32(_Mobi.mobiType);
    return _MOBIHeader(
      headerLength: u32(_Mobi.headerLength),
      textEncoding: u32(_Mobi.textEncoding),
      firstNonBookIndex: u32(_Mobi.firstNonBookIndex),
      fullNameOffset: u32(_Mobi.fullNameOffset),
      fullNameLength: u32(_Mobi.fullNameLength),
      firstImageRecord: u32(_Mobi.firstImageRecord),
      exthFlags: u32(_Mobi.exthFlags),
      extraDataFlags: rec0.length >= _Mobi.extraDataFlags + 2
          ? bd.getUint16(_Mobi.extraDataFlags, Endian.big)
          : 0,
      // 248 / 257 are the Mobipocket type values for KF8-formatted files.
      isKF8: mobiType == 248 || mobiType == 257,
    );
  }
}

// Offsets are absolute inside record 0 (i.e. they include the 16-byte
// PalmDOC header preceding the MOBI magic).
abstract class _Mobi {
  static const minLength = 24;
  static const magic = 16;
  static const headerLength = 20;
  static const mobiType = 24;
  static const textEncoding = 28;
  static const firstNonBookIndex = 80;
  static const fullNameOffset = 84;
  static const fullNameLength = 88;
  static const firstImageRecord = 108;
  static const exthFlags = 128;
  static const extraDataFlags = 242;
}

String? _readFullName(Uint8List rec0, _MOBIHeader mobi) {
  final start = mobi.fullNameOffset;
  final len = mobi.fullNameLength;
  if (start == 0 || len == 0 || start + len > rec0.length) return null;
  return _decodeBytes(rec0.sublist(start, start + len), mobi.textEncoding);
}

class _EXTHData {
  _EXTHData({this.author, this.updatedTitle, this.coverIndex});
  final String? author;
  final String? updatedTitle;
  final int? coverIndex;

  static _EXTHData parse(Uint8List rec0, _MOBIHeader mobi) {
    if (!mobi.hasEXTH) return _EXTHData();
    final base = 16 + mobi.headerLength;
    if (base + 12 > rec0.length) return _EXTHData();
    if (latin1.decode(rec0.sublist(base, base + 4)) != 'EXTH') {
      return _EXTHData();
    }

    final bd = ByteData.sublistView(rec0);
    final count = bd.getUint32(base + 8, Endian.big);
    var cursor = base + 12;
    String? author;
    String? updatedTitle;
    int? coverIndex;

    for (var i = 0; i < count; i++) {
      if (cursor + 8 > rec0.length) break;
      final type = bd.getUint32(cursor, Endian.big);
      final length = bd.getUint32(cursor + 4, Endian.big);
      if (length < 8 || cursor + length > rec0.length) break;
      final data = rec0.sublist(cursor + 8, cursor + length);
      switch (type) {
        case 100:
          author ??= _decodeBytes(data, mobi.textEncoding).trim();
        case 201:
          if (data.length >= 4) {
            coverIndex ??= ByteData.sublistView(data).getUint32(0, Endian.big);
          }
        case 503:
          updatedTitle ??= _decodeBytes(data, mobi.textEncoding).trim();
      }
      cursor += length;
    }
    return _EXTHData(
      author: author,
      updatedTitle: updatedTitle,
      coverIndex: coverIndex,
    );
  }
}

// Each MOBI text record may carry trailing data (multibyte-char overflow,
// index records, ...) appended after the compressed payload. Each bit
// of extraDataFlags above bit 0 marks a variable-length trailer to peel
// off, and bit 0 marks 1-3 multibyte-char overflow bytes whose count is
// encoded in the low 2 bits of the final byte. Get this wrong and
// PalmDOC decompression desynchronises mid-record.
Uint8List _stripTrailers(Uint8List record, int flags) {
  var end = record.length;
  for (var bit = 0x8000; bit > 1; bit >>>= 1) {
    if ((flags & bit) == 0) continue;
    end = _stripVlen(record, end);
    if (end <= 0) return Uint8List(0);
  }
  if ((flags & 1) != 0 && end > 0) {
    end -= (record[end - 1] & 0x3) + 1;
    if (end < 0) end = 0;
  }
  return record.sublist(0, end);
}

int _stripVlen(Uint8List record, int end) {
  var len = 0;
  for (var i = 0; i < 4 && end - 1 - i >= 0; i++) {
    final b = record[end - 1 - i];
    len = (len << 7) | (b & 0x7F);
    if ((b & 0x80) != 0) break;
  }
  return (end - len).clamp(0, end);
}

// PalmDOC compression: a byte-oriented LZ77 variant. Token byte meanings:
//   0x00         literal NUL
//   0x01..0x08   N literal bytes follow
//   0x09..0x7F   literal ASCII byte
//   0x80..0xBF   back-reference: top 2 bits "10", remaining 14 bits hold
//                an 11-bit distance and a 3-bit length-3 in the second byte
//   0xC0..0xFF   ' ' + (b XOR 0x80)
Uint8List _palmdocInflate(Uint8List input) {
  final out = BytesBuilder();
  var i = 0;
  while (i < input.length) {
    final b = input[i++];
    if (b == 0) {
      out.addByte(0);
    } else if (b <= 0x08) {
      final n = b;
      if (i + n > input.length) break;
      for (var k = 0; k < n; k++) {
        out.addByte(input[i + k]);
      }
      i += n;
    } else if (b <= 0x7F) {
      out.addByte(b);
    } else if (b <= 0xBF) {
      if (i >= input.length) break;
      final pair = ((b << 8) | input[i++]) & 0x3FFF;
      final distance = pair >> 3;
      final length = (pair & 0x7) + 3;
      final cur = out.toBytes();
      if (distance == 0 || distance > cur.length) continue;
      for (var k = 0; k < length; k++) {
        out.addByte(out.toBytes()[cur.length + k - distance]);
      }
    } else {
      out.addByte(0x20);
      out.addByte(b ^ 0x80);
    }
  }
  return out.toBytes();
}

Uint8List? _imageAt(_PalmDB db, _MOBIHeader mobi, int coverIndex) {
  if (mobi.firstImageRecord == 0) return null;
  final rec = mobi.firstImageRecord + coverIndex;
  if (rec < 0 || rec >= db.recordCount) return null;
  final bytes = db.recordBytes(rec);
  return _isImage(bytes) ? bytes : null;
}

Uint8List? _firstImage(_PalmDB db, _MOBIHeader mobi) {
  if (mobi.firstImageRecord == 0) return null;
  final stop = (mobi.firstImageRecord + 8).clamp(0, db.recordCount).toInt();
  for (var i = mobi.firstImageRecord; i < stop; i++) {
    final bytes = db.recordBytes(i);
    if (_isImage(bytes)) return bytes;
  }
  return null;
}

bool _isImage(Uint8List b) {
  if (b.length < 4) return false;
  if (b[0] == 0xFF && b[1] == 0xD8) return true; // JPEG
  if (b[0] == 0x89 && b[1] == 0x50 && b[2] == 0x4E && b[3] == 0x47) {
    return true; // PNG
  }
  if (b[0] == 0x47 && b[1] == 0x49 && b[2] == 0x46) return true; // GIF
  return false;
}

// MOBI 6 has no TOC-anchor table like EPUB, so we lean on the publisher's
// own h1/h2 markup. If a book uses no headings we keep it as one segment
// rather than guessing.
final RegExp _headingRe =
    RegExp(r'<h([12])\b[^>]*>([\s\S]*?)</h\1>', caseSensitive: false);

List<TextSegment> _splitOnHeadings(String html) {
  final matches = _headingRe.allMatches(html).toList();
  if (matches.isEmpty) {
    final plain = stripHTML(html);
    return plain.isEmpty
        ? const []
        : [TextSegment(id: 'mobi-1', title: null, text: plain)];
  }

  final out = <TextSegment>[];
  if (matches.first.start > 0) {
    final pre = stripHTML(html.substring(0, matches.first.start));
    if (pre.length >= 200) {
      out.add(TextSegment(id: 'mobi-pre', title: null, text: pre));
    }
  }
  for (var i = 0; i < matches.length; i++) {
    final m = matches[i];
    final next = i + 1 < matches.length ? matches[i + 1].start : html.length;
    final plain = stripHTML(html.substring(m.start, next));
    if (plain.isEmpty) continue;
    final title = stripHTML(m.group(2) ?? '').trim();
    out.add(TextSegment(
      id: 'mobi-${i + 1}',
      title: title.isEmpty ? null : title,
      text: plain,
    ));
  }
  return out;
}

File: InkAndEchoCore/Sources/InkAndEchoCore/Import/MOBIImporter.swift
import Foundation

// Pure-Swift MOBI 6 / PalmDOC importer. Mirrors the in-tree Dart parser at
// xplatform/lib/import/mobi_importer.dart so the iOS / macOS surface gets
// the same coverage. App Store-safe: no subprocess, no GPL dependency.
//
// References:
//   - https://wiki.mobileread.com/wiki/MOBI
//   - https://wiki.mobileread.com/wiki/PDB
//
// KF8 (.azw3) and HUFF/CDIC compressed payloads throw rather than emitting
// garbage; the surface UI prompts the user to convert with Calibre on a
// machine that has it.

public struct MOBIImporter: EBookImporter {
    public init() {}

    public func importBook(from url: URL) async throws -> ImportedBook {
        let data = try Data(contentsOf: url)
        return try decode(data)
    }

    func decode(_ data: Data) throws -> ImportedBook {
        let db = try PalmDB.parse(data)
        guard db.type == "BOOK", db.creator == "MOBI" else {
            throw ImporterError.malformedMOBI(
                #"Not a MOBI file (PDB type="\#(db.type)", creator="\#(db.creator)")."#
            )
        }
        guard db.recordCount > 0 else {
            throw ImporterError.malformedMOBI("MOBI file has no records.")
        }

        let rec0 = db.recordBytes(0)
        let pdoc = try PalmDOCHeader.parse(rec0)
        if pdoc.encryption != 0 {
            throw ImporterError.drmProtected
        }
        let mobi = try MOBIHeader.parse(rec0)
        if pdoc.compression == MOBIConstants.huffCdicCompression {
            throw ImporterError.malformedMOBI(
                "HUFF/CDIC compression unsupported. Convert to EPUB with Calibre."
            )
        }
        if mobi.isKF8 {
            throw ImporterError.unsupportedKF8
        }

        let html = MOBITextReader.read(db: db, pdoc: pdoc, mobi: mobi)
        let exth = EXTHData.parse(rec0: rec0, mobi: mobi)
        let candidate = (exth.updatedTitle ?? MOBITextReader.readFullName(rec0: rec0, mobi: mobi) ?? db.name)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let title = candidate.isEmpty ? db.name : candidate
        let cover = MOBICover.find(db: db, mobi: mobi, exth: exth)

        return ImportedBook(
            title: title,
            author: exth.author ?? "Unknown",
            coverImageData: cover,
            segments: MOBIChapterSplitter.split(html: html),
            totalPages: 0
        )
    }
}

// MARK: - Constants

enum MOBIConstants {
    static let palmDocCompression = 2
    static let huffCdicCompression = 17480
}

// MARK: - PalmDB

struct PalmDB {
    let name: String
    let type: String
    let creator: String
    let recordOffsets: [Int]
    let bytes: Data

    var recordCount: Int { recordOffsets.count }

    static func parse(_ bytes: Data) throws -> PalmDB {
        guard bytes.count >= MOBIBytes.pdbHeaderSize else {
            throw ImporterError.malformedMOBI("File too short to be a PalmDB.")
        }
        let count = Int(MOBIBytes.readUInt16BE(bytes, at: MOBIBytes.pdbRecordCount))
        guard MOBIBytes.pdbHeaderSize + count * MOBIBytes.pdbRecordEntrySize <= bytes.count else {
            throw ImporterError.malformedMOBI("PalmDB record table truncated.")
        }
        var offsets: [Int] = []
        offsets.reserveCapacity(count)
        for i in 0..<count {
            let off = Int(MOBIBytes.readUInt32BE(
                bytes,
                at: MOBIBytes.pdbHeaderSize + i * MOBIBytes.pdbRecordEntrySize
            ))
            offsets.append(off)
        }
        return PalmDB(
            name: MOBIBytes.readCString(bytes, offset: MOBIBytes.pdbName, maxLen: MOBIBytes.pdbNameLength),
            type: MOBIBytes.latin1(bytes, range: MOBIBytes.pdbType..<(MOBIBytes.pdbType + 4)),
            creator: MOBIBytes.latin1(bytes, range: MOBIBytes.pdbCreator..<(MOBIBytes.pdbCreator + 4)),
            recordOffsets: offsets,
            bytes: bytes
        )
    }

    func recordBytes(_ idx: Int) -> Data {
        let start = recordOffsets[idx]
        let end = idx + 1 < recordOffsets.count ? recordOffsets[idx + 1] : bytes.count
        return bytes.subdata(in: start..<end)
    }
}

// MARK: - PalmDOC + MOBI headers

struct PalmDOCHeader {
    let compression: Int
    let recordCount: Int
    let encryption: Int

    static func parse(_ rec0: Data) throws -> PalmDOCHeader {
        guard rec0.count >= 16 else {
            throw ImporterError.malformedMOBI("Record 0 too short.")
        }
        return PalmDOCHeader(
            compression: Int(MOBIBytes.readUInt16BE(rec0, at: 0)),
            recordCount: Int(MOBIBytes.readUInt16BE(rec0, at: 8)),
            encryption: Int(MOBIBytes.readUInt16BE(rec0, at: 12))
        )
    }
}

struct MOBIHeader {
    let headerLength: Int
    let textEncoding: Int
    let firstNonBookIndex: Int
    let fullNameOffset: Int
    let fullNameLength: Int
    let firstImageRecord: Int
    let exthFlags: Int
    let extraDataFlags: Int
    let isKF8: Bool

    var hasEXTH: Bool { (exthFlags & 0x40) != 0 }

    static func parse(_ rec0: Data) throws -> MOBIHeader {
        guard rec0.count >= MOBIBytes.mobiMinLength,
              MOBIBytes.latin1(rec0, range: MOBIBytes.mobiMagic..<(MOBIBytes.mobiMagic + 4)) == "MOBI"
        else {
            throw ImporterError.malformedMOBI("MOBI header not found in record 0.")
        }
        func u32(_ o: Int) -> Int {
            o + 4 <= rec0.count ? Int(MOBIBytes.readUInt32BE(rec0, at: o)) : 0
        }
        let mobiType = u32(MOBIBytes.mobiType)
        let extra = rec0.count >= MOBIBytes.mobiExtraDataFlags + 2
            ? Int(MOBIBytes.readUInt16BE(rec0, at: MOBIBytes.mobiExtraDataFlags))
            : 0
        return MOBIHeader(
            headerLength: u32(MOBIBytes.mobiHeaderLength),
            textEncoding: u32(MOBIBytes.mobiTextEncoding),
            firstNonBookIndex: u32(MOBIBytes.mobiFirstNonBookIndex),
            fullNameOffset: u32(MOBIBytes.mobiFullNameOffset),
            fullNameLength: u32(MOBIBytes.mobiFullNameLength),
            firstImageRecord: u32(MOBIBytes.mobiFirstImageRecord),
            exthFlags: u32(MOBIBytes.mobiExthFlags),
            extraDataFlags: extra,
            // 248 / 257 are the Mobipocket type values for KF8-formatted files.
            isKF8: mobiType == 248 || mobiType == 257
        )
    }
}

// MARK: - EXTH metadata

struct EXTHData {
    let author: String?
    let updatedTitle: String?
    let coverIndex: Int?

    static func parse(rec0: Data, mobi: MOBIHeader) -> EXTHData {
        guard mobi.hasEXTH else { return EXTHData(author: nil, updatedTitle: nil, coverIndex: nil) }
        let base = 16 + mobi.headerLength
        guard base + 12 <= rec0.count,
              MOBIBytes.latin1(rec0, range: base..<(base + 4)) == "EXTH"
        else {
            return EXTHData(author: nil, updatedTitle: nil, coverIndex: nil)
        }

        let count = Int(MOBIBytes.readUInt32BE(rec0, at: base + 8))
        var cursor = base + 12
        var author: String?
        var updatedTitle: String?
        var coverIndex: Int?

        for _ in 0..<count {
            guard cursor + 8 <= rec0.count else { break }
            let type = Int(MOBIBytes.readUInt32BE(rec0, at: cursor))
            let length = Int(MOBIBytes.readUInt32BE(rec0, at: cursor + 4))
            guard length >= 8, cursor + length <= rec0.count else { break }
            let data = rec0.subdata(in: (cursor + 8)..<(cursor + length))
            switch type {
            case 100:
                if author == nil {
                    author = MOBIBytes.decode(data, encoding: mobi.textEncoding)
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                }
            case 201:
                if coverIndex == nil, data.count >= 4 {
                    coverIndex = Int(MOBIBytes.readUInt32BE(data, at: 0))
                }
            case 503:
                if updatedTitle == nil {
                    updatedTitle = MOBIBytes.decode(data, encoding: mobi.textEncoding)
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                }
            default:
                break
            }
            cursor += length
        }
        return EXTHData(author: author, updatedTitle: updatedTitle, coverIndex: coverIndex)
    }
}

// MARK: - Text assembly + PalmDOC decompression

enum MOBITextReader {
    static func read(db: PalmDB, pdoc: PalmDOCHeader, mobi: MOBIHeader) -> String {
        let lastText = mobi.firstNonBookIndex > 1 ? mobi.firstNonBookIndex - 1 : pdoc.recordCount
        var buf = Data()
        let upper = min(lastText, db.recordCount - 1)
        if upper >= 1 {
            for i in 1...upper {
                let raw = stripTrailers(db.recordBytes(i), flags: mobi.extraDataFlags)
                if pdoc.compression == MOBIConstants.palmDocCompression {
                    buf.append(palmdocInflate(raw))
                } else {
                    buf.append(raw)
                }
            }
        }
        return MOBIBytes.decode(buf, encoding: mobi.textEncoding)
    }

    static func readFullName(rec0: Data, mobi: MOBIHeader) -> String? {
        let start = mobi.fullNameOffset
        let len = mobi.fullNameLength
        guard start > 0, len > 0, start + len <= rec0.count else { return nil }
        return MOBIBytes.decode(rec0.subdata(in: start..<(start + len)), encoding: mobi.textEncoding)
    }

    // Each MOBI text record may carry trailing data (multibyte-char overflow,
    // index records, ...) appended after the compressed payload. Each bit of
    // extraDataFlags above bit 0 marks a variable-length trailer to peel off,
    // and bit 0 marks 1-3 multibyte-char overflow bytes whose count is encoded
    // in the low 2 bits of the final byte. Get this wrong and PalmDOC
    // decompression desynchronises mid-record.
    static func stripTrailers(_ record: Data, flags: Int) -> Data {
        var end = record.count
        var bit = 0x8000
        while bit > 1 {
            defer { bit >>= 1 }
            if flags & bit == 0 { continue }
            end = stripVlen(record, end: end)
            if end <= 0 { return Data() }
        }
        if flags & 1 != 0, end > 0 {
            let n = Int(record[end - 1] & 0x3) + 1
            end -= n
            if end < 0 { end = 0 }
        }
        return record.subdata(in: 0..<end)
    }

    static func stripVlen(_ record: Data, end: Int) -> Int {
        var len = 0
        var i = 0
        while i < 4, end - 1 - i >= 0 {
            let b = Int(record[end - 1 - i])
            len = (len << 7) | (b & 0x7F)
            if b & 0x80 != 0 { break }
            i += 1
        }
        let newEnd = end - len
        return newEnd < 0 ? 0 : (newEnd > end ? end : newEnd)
    }

    // PalmDOC compression: a byte-oriented LZ77 variant. Token byte meanings:
    //   0x00         literal NUL
    //   0x01..0x08   N literal bytes follow
    //   0x09..0x7F   literal ASCII byte
    //   0x80..0xBF   back-reference: top 2 bits "10", remaining 14 bits hold
    //                an 11-bit distance and a 3-bit length-3 in the second byte
    //   0xC0..0xFF   ' ' + (b XOR 0x80)
    static func palmdocInflate(_ input: Data) -> Data {
        var out = Data()
        var i = 0
        while i < input.count {
            let b = input[i]
            i += 1
            if b == 0 {
                out.append(0)
            } else if b <= 0x08 {
                let n = Int(b)
                if i + n > input.count { break }
                out.append(input.subdata(in: i..<(i + n)))
                i += n
            } else if b <= 0x7F {
                out.append(b)
            } else if b <= 0xBF {
                if i >= input.count { break }
                let pair = ((Int(b) << 8) | Int(input[i])) & 0x3FFF
                i += 1
                let distance = pair >> 3
                let length = (pair & 0x7) + 3
                let cur = out.count
                if distance == 0 || distance > cur { continue }
                for k in 0..<length {
                    out.append(out[cur + k - distance])
                }
            } else {
                out.append(0x20)
                out.append(b ^ 0x80)
            }
        }
        return out
    }
}

// MARK: - Cover

enum MOBICover {
    static func find(db: PalmDB, mobi: MOBIHeader, exth: EXTHData) -> Data? {
        if let idx = exth.coverIndex, let bytes = imageAt(db: db, mobi: mobi, coverIndex: idx) {
            return bytes
        }
        return firstImage(db: db, mobi: mobi)
    }

    static func imageAt(db: PalmDB, mobi: MOBIHeader, coverIndex: Int) -> Data? {
        guard mobi.firstImageRecord > 0 else { return nil }
        let rec = mobi.firstImageRecord + coverIndex
        guard rec >= 0, rec < db.recordCount else { return nil }
        let bytes = db.recordBytes(rec)
        return isImage(bytes) ? bytes : nil
    }

    static func firstImage(db: PalmDB, mobi: MOBIHeader) -> Data? {
        guard mobi.firstImageRecord > 0 else { return nil }
        let stop = min(mobi.firstImageRecord + 8, db.recordCount)
        if stop <= mobi.firstImageRecord { return nil }
        for i in mobi.firstImageRecord..<stop {
            let bytes = db.recordBytes(i)
            if isImage(bytes) { return bytes }
        }
        return nil
    }

    static func isImage(_ b: Data) -> Bool {
        guard b.count >= 4 else { return false }
        if b[0] == 0xFF, b[1] == 0xD8 { return true } // JPEG
        if b[0] == 0x89, b[1] == 0x50, b[2] == 0x4E, b[3] == 0x47 { return true } // PNG
        if b[0] == 0x47, b[1] == 0x49, b[2] == 0x46 { return true } // GIF
        return false
    }
}

// MARK: - Chapter splitting

// MOBI 6 has no TOC-anchor table like EPUB, so we lean on the publisher's
// own h1/h2 markup. If a book uses no headings we keep it as one segment
// rather than guessing.
enum MOBIChapterSplitter {
    private static let headingRegex: NSRegularExpression = {
        try! NSRegularExpression(
            pattern: "<h([12])\\b[^>]*>([\\s\\S]*?)</h\\1>",
            options: .caseInsensitive
        )
    }()

    static func split(html: String) -> [TextSegment] {
        let nsRange = NSRange(html.startIndex..<html.endIndex, in: html)
        let matches = headingRegex.matches(in: html, range: nsRange)
        if matches.isEmpty {
            let plain = stripHTML(html)
            return plain.isEmpty
                ? []
                : [TextSegment(id: "mobi-1", title: nil, text: plain)]
        }

        var out: [TextSegment] = []
        if let firstRange = Range(matches[0].range, in: html), firstRange.lowerBound > html.startIndex {
            let pre = stripHTML(String(html[html.startIndex..<firstRange.lowerBound]))
            if pre.count >= 200 {
                out.append(TextSegment(id: "mobi-pre", title: nil, text: pre))
            }
        }
        for (i, m) in matches.enumerated() {
            guard let mRange = Range(m.range, in: html) else { continue }
            let endIdx = i + 1 < matches.count
                ? Range(matches[i + 1].range, in: html)?.lowerBound ?? html.endIndex
                : html.endIndex
            let plain = stripHTML(String(html[mRange.lowerBound..<endIdx]))
            if plain.isEmpty { continue }
            var title: String? = nil
            if m.numberOfRanges > 2, let titleRange = Range(m.range(at: 2), in: html) {
                let t = stripHTML(String(html[titleRange]))
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                title = t.isEmpty ? nil : t
            }
            out.append(TextSegment(id: "mobi-\(i + 1)", title: title, text: plain))
        }
        return out
    }
}

// MARK: - Byte / string helpers

enum MOBIBytes {
    // PalmDB layout
    static let pdbHeaderSize = 78
    static let pdbRecordEntrySize = 8
    static let pdbName = 0
    static let pdbNameLength = 32
    static let pdbType = 60
    static let pdbCreator = 64
    static let pdbRecordCount = 76

    // MOBI header offsets (absolute inside record 0; include the 16-byte
    // PalmDOC header preceding the MOBI magic).
    static let mobiMinLength = 24
    static let mobiMagic = 16
    static let mobiHeaderLength = 20
    static let mobiType = 24
    static let mobiTextEncoding = 28
    static let mobiFirstNonBookIndex = 80
    static let mobiFullNameOffset = 84
    static let mobiFullNameLength = 88
    static let mobiFirstImageRecord = 108
    static let mobiExthFlags = 128
    static let mobiExtraDataFlags = 242

    static func readUInt16BE(_ data: Data, at offset: Int) -> UInt16 {
        let hi = UInt16(data[offset])
        let lo = UInt16(data[offset + 1])
        return (hi << 8) | lo
    }

    static func readUInt32BE(_ data: Data, at offset: Int) -> UInt32 {
        let b0 = UInt32(data[offset])
        let b1 = UInt32(data[offset + 1])
        let b2 = UInt32(data[offset + 2])
        let b3 = UInt32(data[offset + 3])
        return (b0 << 24) | (b1 << 16) | (b2 << 8) | b3
    }

    static func latin1(_ data: Data, range: Range<Int>) -> String {
        let slice = data.subdata(in: range)
        return String(data: slice, encoding: .isoLatin1) ?? ""
    }

    static func readCString(_ data: Data, offset: Int, maxLen: Int) -> String {
        var stop = offset + maxLen
        var i = offset
        while i < offset + maxLen, i < data.count {
            if data[i] == 0 { stop = i; break }
            i += 1
        }
        return latin1(data, range: offset..<stop)
    }

    static func decode(_ data: Data, encoding: Int) -> String {
        // 65001 = UTF-8; 1252 (and absent) = Windows-1252 / latin1 fallback.
        if encoding == 65001 {
            return String(data: data, encoding: .utf8) ?? String(data: data, encoding: .isoLatin1) ?? ""
        }
        return String(data: data, encoding: .windowsCP1252)
            ?? String(data: data, encoding: .isoLatin1)
            ?? ""
    }
}
